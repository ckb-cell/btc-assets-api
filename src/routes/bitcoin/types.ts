import { z } from 'zod';

export const ChainInfo = z.object({
  chain: z.string(),
  blocks: z.number(),
  bestblockhash: z.string(),
  difficulty: z.number(),
  mediantime: z.number(),
});

export const Block = z.object({
  id: z.string(),
  height: z.number(),
  version: z.number(),
  timestamp: z.number(),
  tx_count: z.number(),
  size: z.number(),
  weight: z.number(),
  merkle_root: z.string(),
  previousblockhash: z.string(),
  mediantime: z.number(),
  nonce: z.number(),
  bits: z.number(),
  difficulty: z.number(),
});

export const Status = z.object({
  confirmed: z.boolean(),
  block_height: z.number().optional(),
  block_hash: z.string().optional(),
  block_time: z.number().optional(),
});

export const Balance = z.object({
  address: z.string(),
  total_satoshi: z.number().describe('Total balance in satoshi (available + pending + dust/rgbpp-bound)'),
  pending_satoshi: z.number().describe('Pending balance in satoshi (unconfirmed)'),
  satoshi: z.number().describe('@deprecated Use available_satoshi'),
  available_satoshi: z.number().describe('Available balance in satoshi (confirmed and not dust/rgbpp-bound)'),
  dust_satoshi: z.number().describe('Dust balance in satoshi (confirmed and below min_satoshi threshold)'),
  rgbpp_satoshi: z.number().describe('RGB++ bound balance in satoshi (confirmed and RGB++ bound)'),
  utxo_count: z.number(),
});

export const UTXO = z.object({
  txid: z.string(),
  vout: z.number(),
  value: z.number(),
  status: Status,
});

const Output = z.object({
  scriptpubkey: z.string(),
  scriptpubkey_asm: z.string(),
  scriptpubkey_type: z.string(),
  scriptpubkey_address: z.string().optional(),
  value: z.number(),
});

const Input = z.object({
  txid: z.string(),
  vout: z.number(),
  prevout: Output.or(z.null()),
  scriptsig: z.string(),
  scriptsig_asm: z.string(),
  witness: z.array(z.string()).optional(),
  is_coinbase: z.boolean(),
  sequence: z.coerce.number(),
});

export const Transaction = z.object({
  txid: z.string(),
  version: z.number(),
  locktime: z.number(),
  vin: z.array(Input),
  vout: z.array(Output),
  size: z.number(),
  weight: z.number(),
  fee: z.number(),
  status: Status,
});

export const RecommendedFees = z.object({
  fastestFee: z.number(),
  halfHourFee: z.number(),
  hourFee: z.number(),
  economyFee: z.number(),
  minimumFee: z.number(),
});

export type ChainInfo = z.infer<typeof ChainInfo>;
export type Block = z.infer<typeof Block>;
export type Balance = z.infer<typeof Balance>;
export type UTXO = z.infer<typeof UTXO>;
export type Transaction = z.infer<typeof Transaction>;
export type RecommendedFees = z.infer<typeof RecommendedFees>;
