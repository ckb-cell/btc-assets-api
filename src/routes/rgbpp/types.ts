import z from 'zod';

export const Script = z.object({
  codeHash: z.string(),
  args: z.string(),
  hashType: z.enum(['type', 'data', 'data1', 'data2']),
});
export type Script = z.infer<typeof Script>;

export const CellDep = z.object({
  outPoint: z
    .object({
      txHash: z.string(),
      index: z.string(),
    })
    .or(z.null()),
  depType: z.enum(['depGroup', 'code']),
});
export type CellDep = z.infer<typeof CellDep>;

export const InputCell = z.object({
  previousOutput: z
    .object({
      txHash: z.string(),
      index: z.string(),
    })
    .or(z.null()),
  since: z.string(),
});
export type InputCell = z.infer<typeof InputCell>;

export const OutputCell = z.object({
  capacity: z.string(),
  lock: Script,
  type: Script.or(z.null()).optional(),
});
export type OutputCell = z.infer<typeof OutputCell>;

export const Cell = z.object({
  cellOutput: OutputCell,
  data: z.string(),
  outPoint: z
    .object({
      txHash: z.string(),
      index: z.string(),
    })
    .or(z.null())
    .optional(),
  blockHash: z.string().optional(),
  blockNumber: z.string().optional(),
  txIndex: z.string().optional(),
});
export type Cell = z.infer<typeof Cell>;

export const CKBRawTransaction = z.object({
  version: z.string(),
  cellDeps: z.array(CellDep),
  headerDeps: z.array(z.string()),
  inputs: z.array(InputCell),
  outputs: z.array(OutputCell),
  outputsData: z.array(z.string()),
  witnesses: z.array(z.string()).default([]),
});
export type CKBRawTransaction = z.infer<typeof CKBRawTransaction>;

export const CKBTransaction = z.object({
  cellDeps: z.array(CellDep),
  inputs: z.array(InputCell),
  outputs: z.array(OutputCell),
  outputsData: z.array(z.string()),
  headerDeps: z.array(z.string()),
  hash: z.string(),
  version: z.string(),
  witnesses: z.array(z.string()),
});
export type CKBTransaction = z.infer<typeof CKBTransaction>;

export const CKBVirtualResult = z.object({
  ckbRawTx: CKBRawTransaction,
  commitment: z.string(),
  needPaymasterCell: z.boolean(),
  sumInputsCapacity: z.string(),
});
export type CKBVirtualResult = z.infer<typeof CKBVirtualResult>;

export const XUDTBalance = z.object({
  name: z.string(),
  decimal: z.number(),
  symbol: z.string(),
  amount: z.string(),
  typeHash: z.string(),
});
export type XUDTBalance = z.infer<typeof XUDTBalance>;
