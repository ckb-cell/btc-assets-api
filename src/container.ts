import { createContainer, InjectionMode, asValue, asClass, asFunction } from 'awilix';
import { Redis } from 'ioredis';
import pino from 'pino';
import Bitcoind from './services/bitcoind';
import ElectrsAPI from './services/electrs';
import { env } from './env';
import TransactionManager from './services/transaction';
import Paymaster from './services/paymaster';
import { RPC as CkbRPC, Indexer as CkbIndexer } from '@ckb-lumos/lumos';
import Unlocker from './services/unlocker';

export interface Cradle {
  env: typeof env;
  logger: pino.BaseLogger;
  redis: Redis;
  ckbRpc: CkbRPC;
  ckbIndexer: CkbIndexer;
  bitcoind: Bitcoind;
  electrs: ElectrsAPI;
  paymaster: Paymaster;
  unlocker: Unlocker;
  transactionManager: TransactionManager;
}

const container = createContainer<Cradle>({
  injectionMode: InjectionMode.PROXY,
  strict: true,
});

container.register({
  env: asValue(env),
  logger: asValue(pino()),
  redis: asValue(
    new Redis(env.REDIS_URL, {
      maxRetriesPerRequest: null,
    }),
  ),
  ckbRpc: asFunction(() => new CkbRPC(env.CKB_RPC_URL)).singleton(),
  ckbIndexer: asFunction(() => new CkbIndexer(env.CKB_INDEXER_URL)).singleton(),
  bitcoind: asClass(Bitcoind).singleton(),
  electrs: asClass(ElectrsAPI).singleton(),
  paymaster: asClass(Paymaster).singleton(),
  transactionManager: asClass(TransactionManager).singleton(),
  unlocker: asClass(Unlocker).singleton(),
});

export default container;
