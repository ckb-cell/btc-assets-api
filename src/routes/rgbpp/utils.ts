import { bytes, createBytesCodec, createFixedBytesCodec, molecule } from '@ckb-lumos/codec';

const { table } = molecule;

const fallbackBytesCodec = createBytesCodec({
  pack: bytes.bytify,
  unpack: bytes.hexify,
});

function createFallbackFixedBytesCodec(byteLength: number) {
  return createFixedBytesCodec({
    pack: bytes.bytify,
    unpack: bytes.hexify,
    byteLength,
  });
}

const byte = createFallbackFixedBytesCodec(1);
export const Bytes = fallbackBytesCodec;
export const Byte32 = createFallbackFixedBytesCodec(32);

export const TypeScript = table(
  {
    codeHash: Byte32,
    hashType: byte,
    args: Bytes,
  },
  ['codeHash', 'hashType', 'args'],
);
