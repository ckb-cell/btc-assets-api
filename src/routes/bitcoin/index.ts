import { FastifyPluginCallback } from 'fastify';
import { Server } from 'http';
import infoRoute from './info';
import blockRoutes from './block';
import transactionRoutes from './transaction';
import addressRoutes from './address';
import container from '../../container';
import { ZodTypeProvider } from 'fastify-type-provider-zod';
import BitcoinClient from '../../services/bitcoin';
import feesRoutes from './fees';

const bitcoinRoutes: FastifyPluginCallback<Record<never, never>, Server, ZodTypeProvider> = (fastify, _, done) => {
  fastify.decorate('bitcoin', container.resolve<BitcoinClient>('bitcoin'));

  fastify.register(infoRoute);
  fastify.register(blockRoutes, { prefix: '/block' });
  fastify.register(transactionRoutes, { prefix: '/transaction' });
  fastify.register(addressRoutes, { prefix: '/address' });
  fastify.register(feesRoutes, { prefix: '/fees' });
  done();
};

export default bitcoinRoutes;
