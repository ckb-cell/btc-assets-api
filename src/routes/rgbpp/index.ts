import { FastifyPluginCallback } from 'fastify';
import { Server } from 'http';
import container from '../../container';
import transactionRoutes from './transaction';
import { ZodTypeProvider } from 'fastify-type-provider-zod';

const rgbppRoutes: FastifyPluginCallback<Record<never, never>, Server, ZodTypeProvider> = (fastify, _, done) => {
  fastify.decorate('transactionManager', container.resolve('transactionManager'));
  fastify.register(transactionRoutes, { prefix: '/transaction' });
  done();
};

export default rgbppRoutes;
