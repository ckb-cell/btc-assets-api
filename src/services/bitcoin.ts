import mempoolJS from '@mempool/mempool.js';
import { Cradle } from '../container';
import { Block, ChainInfo, Transaction, UTXO } from '../routes/bitcoin/types';
import { NetworkType } from '../constants';
import { AxiosError } from 'axios';
import Electrs from '../utils/electrs';
import * as Sentry from '@sentry/node';

// https://github.com/mempool/electrs/blob/d4f788fc3d7a2b4eca4c5629270e46baba7d0f19/src/errors.rs#L6
export enum MempoolErrorMessage {
  Connection = 'Connection error',
  Interrupt = 'Interruption by external signal',
  TooManyUtxos = 'Too many unspent transaction outputs',
  TooManyTxs = 'Too many history transactions',
  ElectrumClient = 'Electrum client error',
}

export enum BitcoinClientErrorCode {
  Connection = 0x1000, // 4096
  Interrupt = 0x1001, // 4097
  TooManyUtxos = 0x1002, // 4098
  TooManyTxs = 0x1003, // 4099
  ElectrumClient = 0x1004, // 4100

  MempoolUnknown = 0x1111, // 4369
}

const BitcoinClientErrorMap = {
  [MempoolErrorMessage.Connection]: BitcoinClientErrorCode.Connection,
  [MempoolErrorMessage.Interrupt]: BitcoinClientErrorCode.Interrupt,
  [MempoolErrorMessage.TooManyUtxos]: BitcoinClientErrorCode.TooManyUtxos,
  [MempoolErrorMessage.TooManyTxs]: BitcoinClientErrorCode.TooManyTxs,
  [MempoolErrorMessage.ElectrumClient]: BitcoinClientErrorCode.ElectrumClient,
};

export class BitcoinClientAPIError extends Error {
  public statusCode = 500;
  public errorCode: BitcoinClientErrorCode;

  constructor(message: string) {
    super(message);
    this.name = this.constructor.name;

    const errorKey = Object.keys(BitcoinClientErrorMap).find((msg) => message.startsWith(msg));
    this.errorCode = BitcoinClientErrorMap[errorKey as MempoolErrorMessage] ?? BitcoinClientErrorCode.MempoolUnknown;
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const wrapTry = async <T extends (...args: any) => Promise<any>>(fn: T): Promise<ReturnType<T>> => {
  if (typeof fn !== 'function') {
    throw new Error('wrapTry: fn must be a function');
  }

  try {
    const ret = await fn();
    return ret;
  } catch (err) {
    if ((err as AxiosError).isAxiosError) {
      const error = new BitcoinClientAPIError((err as AxiosError).message);
      if ((err as AxiosError).response) {
        error.statusCode = (err as AxiosError).response?.status || 500;
      }
      throw error;
    }
    throw err;
  }
};

export default class BitcoinClient {
  private cradle: Cradle;
  private mempool: ReturnType<typeof mempoolJS>;
  private electrs?: Electrs;

  constructor(cradle: Cradle) {
    this.cradle = cradle;

    const url = new URL(cradle.env.BITCOIN_MEMPOOL_SPACE_API_URL);
    this.mempool = mempoolJS({
      hostname: url.hostname,
      network: cradle.env.NETWORK,
    });

    if (cradle.env.BITCOIN_ELECTRS_API_URL) {
      this.electrs = new Electrs(cradle);
    }
  }

  public async checkNetwork(network: NetworkType) {
    const hash = await this.getBlockHashByHeight(0);
    switch (network) {
      case NetworkType.mainnet:
        // Bitcoin mainnet genesis block hash
        if (hash !== '000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f') {
          throw new Error('Mempool API is not running on mainnet');
        }
        break;
      case NetworkType.testnet:
        // Bitcoin testnet genesis block hash
        if (hash !== '000000000933ea01ad0ee984209779baaec3ced90fa3f408719526f8d77f4943') {
          throw new Error('Mempool API is not running on testnet');
        }
        break;
      default:
    }
  }

  public async getBlockchainInfo(): Promise<ChainInfo> {
    const hash = await this.mempool.bitcoin.blocks.getBlocksTipHash();
    const tip = await this.mempool.bitcoin.blocks.getBlock({ hash });

    const { difficulty, mediantime } = tip;
    return {
      chain: this.cradle.env.NETWORK === 'mainnet' ? 'main' : 'test',
      blocks: tip.height,
      bestblockhash: hash,
      difficulty,
      mediantime,
    };
  }

  public async sendRawTransaction(txhex: string): Promise<string> {
    return wrapTry(async () => {
      try {
        const txid = await this.mempool.bitcoin.transactions.postTx({ txhex });
        return txid as string;
      } catch (err) {
        this.cradle.logger.error(err);
        Sentry.captureException(err);
        if (this.electrs) {
          return this.electrs.sendRawTransaction(txhex);
        }
        throw err;
      }
    });
  }

  public async getUtxoByAddress(address: string): Promise<UTXO[]> {
    return wrapTry(async () => {
      try {
        const utxo = await this.mempool.bitcoin.addresses.getAddressTxsUtxo({ address });
        return utxo.map((utxo) => UTXO.parse(utxo));
      } catch (err) {
        this.cradle.logger.error(err);
        Sentry.captureException(err);
        if (this.electrs) {
          return this.electrs.getUtxoByAddress(address);
        }
        throw err;
      }
    });
  }

  public async getTransactionsByAddress(address: string, after_txid?: string): Promise<Transaction[]> {
    return wrapTry(async () => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error after_txid is not defined in the type definition
      const txs = await this.mempool.bitcoin.addresses.getAddressTxs({ address, after_txid });
      return txs.map((tx) => Transaction.parse(tx));
    });
  }

  public async getTransaction(txid: string): Promise<Transaction> {
    return wrapTry(async () => {
      const tx = await this.mempool.bitcoin.transactions.getTx({ txid });
      return Transaction.parse(tx);
    });
  }

  public async getTransactionHex(txid: string): Promise<string> {
    return wrapTry(() => this.mempool.bitcoin.transactions.getTxHex({ txid }));
  }

  public async getBlockByHash(hash: string): Promise<Block> {
    return wrapTry(async () => {
      const block = await this.mempool.bitcoin.blocks.getBlock({ hash });
      return block;
    });
  }

  public async getBlockByHeight(height: number): Promise<Block> {
    return wrapTry(async () => {
      try {
        const hash = await this.mempool.bitcoin.blocks.getBlockHeight({ height });
        const block = await this.mempool.bitcoin.blocks.getBlock({ hash });
        return Block.parse(block);
      } catch (err) {
        this.cradle.logger.error(err);
        Sentry.captureException(err);
        if (this.electrs) {
          const hash = await this.electrs.getBlockHashByHeight(height);
          const block = await this.electrs.getBlockByHash(hash);
          return Block.parse(block);
        }
        throw err;
      }
    });
  }

  public async getBlockHashByHeight(height: number): Promise<string> {
    return wrapTry(async () => {
      try {
        const hash = await this.mempool.bitcoin.blocks.getBlockHeight({ height });
        return hash;
      } catch (err) {
        this.cradle.logger.error(err);
        Sentry.captureException(err);
        if (this.electrs) {
          const hash = await this.electrs.getBlockHashByHeight(height);
          return hash;
        }
        throw err;
      }
    });
  }

  public async getBlockHeaderByHash(hash: string) {
    return wrapTry(async () => {
      try {
        return this.mempool.bitcoin.blocks.getBlockHeader({ hash });
      } catch (err) {
        this.cradle.logger.error(err);
        Sentry.captureException(err);
        if (this.electrs) {
          return this.electrs.getBlockHeaderByHash(hash);
        }
        throw err;
      }
    });
  }

  public async getBlockTxIdsByHash(hash: string): Promise<string[]> {
    return wrapTry(async () => {
      try {
        const txids = await this.mempool.bitcoin.blocks.getBlockTxids({ hash });
        return txids;
      } catch (err) {
        this.cradle.logger.error(err);
        Sentry.captureException(err);
        if (this.electrs) {
          return this.electrs.getBlockTxIdsByHash(hash);
        }
        throw err;
      }
    });
  }

  public async getTip(): Promise<number> {
    return wrapTry(async () => {
      try {
        return this.mempool.bitcoin.blocks.getBlocksTipHeight();
      } catch (err) {
        this.cradle.logger.error(err);
        Sentry.captureException(err);
        if (this.electrs) {
          return this.electrs.getTip();
        }
        throw err;
      }
    });
  }
}
