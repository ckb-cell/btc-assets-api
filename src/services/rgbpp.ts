import { UTXO } from './bitcoin/schema';
import pLimit from 'p-limit';
import asyncRetry from 'async-retry';
import { Cradle } from '../container';
import { CKBIndexerQueryOptions } from '@ckb-lumos/ckb-indexer/lib/type';
import { buildRgbppLockArgs, genRgbppLockScript } from '@rgbpp-sdk/ckb';
import * as Sentry from '@sentry/node';
import { Script } from '@ckb-lumos/lumos';
import { Job } from 'bullmq';
import { z } from 'zod';
import { Cell } from '../routes/rgbpp/types';
import BaseQueueWorker from './base/queue-worker';
import DataCache from './base/data-cache';

type RgbppUtxoCellsPair = {
  utxo: UTXO;
  cells: Cell[];
};

interface IRgbppCollectRequest {
  btcAddress: string;
  utxos: UTXO[];
}

interface IRgbppCollectJobReturn {
  [key: string]: Cell[];
}

export interface IProcessCallbacks {
  onActive?: (job: Job<IRgbppCollectRequest>) => void;
  onCompleted?: (job: Job<IRgbppCollectRequest>) => void;
  onFailed?: (job: Job<IRgbppCollectRequest> | undefined, err: Error) => void;
}

export const RGBPP_COLLECTOR_QUEUE_NAME = 'rgbpp-collector-queue';

export default class RgbppCollector extends BaseQueueWorker<IRgbppCollectRequest, IRgbppCollectJobReturn> {
  private limit: pLimit.Limit;
  private dataCahe: DataCache<IRgbppCollectJobReturn>;

  constructor(private cradle: Cradle) {
    super({
      name: RGBPP_COLLECTOR_QUEUE_NAME,
      connection: cradle.redis,
      worker: {
        lockDuration: 60_000,
        removeOnComplete: { count: 0 },
        removeOnFail: { count: 0 },
      },
    });
    this.dataCahe = new DataCache(cradle.redis, {
      prefix: 'rgbpp-collector-data',
      schema: z.record(z.array(Cell)),
    });
    this.limit = pLimit(cradle.env.CKB_RPC_MAX_CONCURRENCY);
  }

  /**
   * Get the rgbpp cells from cache
   * @param btcAddress - the btc address
   */
  public async getRgbppCellsFromCache(btcAddress: string) {
    const data = await this.dataCahe.get(btcAddress);
    if (!data) {
      return null;
    }
    return Object.values(data).flat();
  }

  /**
   * Get the cells for the utxo, use ckb indexer to query the cells
   * @param utxo - the utxo to collect
   * @param typeScript - the type script to filter the cells
   */
  private async getRgbppCellsByUtxo(utxo: UTXO, typeScript?: Script): Promise<Cell[]> {
    try {
      const { txid, vout } = utxo;
      const args = buildRgbppLockArgs(vout, txid);

      const query: CKBIndexerQueryOptions = {
        lock: genRgbppLockScript(args, process.env.NETWORK === 'mainnet'),
      };

      if (typeScript) {
        query.type = typeScript;
      }

      const collector = this.cradle.ckb.indexer.collector(query).collect();
      const cells: Cell[] = [];
      for await (const cell of collector) {
        cells.push(cell);
      }
      return cells;
    } catch (e) {
      Sentry.withScope((scope) => {
        scope.captureException(e);
      });
      throw e;
    }
  }

  /**
   * Collect the cells for the utxos, return the utxo and the cells
   * @param utxos - the utxos to collect
   * @param typeScript - the type script to filter the cells
   */
  public async collectRgbppUtxoCellsPairs(utxos: UTXO[], typeScript?: Script): Promise<RgbppUtxoCellsPair[]> {
    const cells = await Promise.all(
      utxos.map((utxo) => {
        return this.limit(() =>
          asyncRetry(
            async () => {
              const cells = await this.getRgbppCellsByUtxo(utxo, typeScript);
              return { utxo, cells };
            },
            { retries: 2 },
          ),
        );
      }),
    );

    return cells.filter(({ cells }) => cells.length > 0);
  }

  /**
   * Enqueue a collect job to the queue
   * @param utxos - the utxos to collect
   */
  public async enqueueCollectJob(
    btcAddress: string,
    utxos: UTXO[],
    allowDuplicate?: boolean,
  ): Promise<Job<IRgbppCollectRequest>> {
    let jobId = btcAddress;
    if (allowDuplicate) {
      jobId = `${btcAddress}:${Date.now()}`;
    }
    return this.addJob(jobId, { btcAddress, utxos });
  }

  /**
   * Process the collect job, collect the cells for the utxos
   * concurrently controlled by the CKB_RPC_MAX_CONCURRENCY
   * retry 2 times if failed, and return the utxo and cells
   */
  public async process(job: Job<IRgbppCollectRequest>) {
    const { btcAddress, utxos } = job.data;
    const pairs = await this.collectRgbppUtxoCellsPairs(utxos);
    const data = pairs.reduce((acc, { utxo, cells }) => {
      const key = `${utxo.txid}:${utxo.vout}`;
      acc[key] = cells;
      return acc;
    }, {} as IRgbppCollectJobReturn);
    this.dataCahe.set(btcAddress, data);
    return data;
  }
}
