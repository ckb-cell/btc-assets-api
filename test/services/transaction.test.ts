import { describe, beforeEach, expect, test, vi } from 'vitest';
import TransactionManager, { ITransactionRequest } from '../../src/services/transaction';
import container from '../../src/container';
import { CKBVirtualResult, InputCell, OutputCell } from '../../src/routes/rgbpp/types';
import { Transaction } from '../../src/routes/bitcoin/types';
import { calculateCommitment } from '@rgbpp-sdk/ckb/lib/utils/rgbpp';

const commitment = calculateCommitment({
  inputs: [] as InputCell[],
  outputs: [] as OutputCell[],
} as CKBVirtualResult['ckbRawTx']);

describe('transactionManager', () => {
  let transactionManager: TransactionManager;
  const cradle = container.cradle;

  beforeEach(async () => {
    transactionManager = new TransactionManager(cradle);
  });

  test('verifyTransaction: should return true for valid transaction', async () => {
    vi.spyOn(
      transactionManager as unknown as {
        getCommitmentFromBtcTx: (txid: string) => Promise<Buffer>;
      },
      'getCommitmentFromBtcTx',
    ).mockResolvedValueOnce(Buffer.from(commitment, 'hex'));

    const transactionRequest: ITransactionRequest = {
      txid: 'bb8c92f11920824db22b379c0ef491dea2d819e721d5df296bebc67a0568ea0f',
      ckbVirtualResult: {
        ckbRawTx: { inputs: [] as InputCell[], outputs: [] as OutputCell[] } as CKBVirtualResult['ckbRawTx'],
        commitment,
        sumInputsCapacity: '1000',
        needPaymasterCell: false,
      },
    };
    // FIXME: mock electrs getTransaction
    const btcTx = await cradle.electrs.getTransaction(transactionRequest.txid);
    const isValid = await transactionManager.verifyTransaction(transactionRequest, btcTx);
    expect(isValid).toBe(true);
  });

  test('verifyTransaction: should return false for mismatch commitment', async () => {
    vi.spyOn(
      transactionManager as unknown as {
        getCommitmentFromBtcTx: (txid: string) => Promise<Buffer>;
      },
      'getCommitmentFromBtcTx',
    ).mockResolvedValueOnce(Buffer.from('mismatchcommitment', 'hex'));

    const transactionRequest: ITransactionRequest = {
      txid: 'bb8c92f11920824db22b379c0ef491dea2d819e721d5df296bebc67a0568ea0f',
      ckbVirtualResult: {
        ckbRawTx: { inputs: [] as InputCell[], outputs: [] as OutputCell[] } as CKBVirtualResult['ckbRawTx'],
        commitment,
        sumInputsCapacity: '1000',
        needPaymasterCell: false,
      },
    };
    // FIXME: mock electrs getTransaction
    const btcTx = await cradle.electrs.getTransaction(transactionRequest.txid);
    const isValid = await transactionManager.verifyTransaction(transactionRequest, btcTx);
    expect(isValid).toBe(false);
  });

  test('verifyTransaction: should return false for mismatch ckb tx', async () => {
    const commitment = 'mismatchcommitment';
    vi.spyOn(
      transactionManager as unknown as {
        getCommitmentFromBtcTx: (txid: string) => Promise<Buffer>;
      },
      'getCommitmentFromBtcTx',
    ).mockResolvedValueOnce(Buffer.from(commitment, 'hex'));

    const transactionRequest: ITransactionRequest = {
      txid: 'bb8c92f11920824db22b379c0ef491dea2d819e721d5df296bebc67a0568ea0f',
      ckbVirtualResult: {
        ckbRawTx: { inputs: [] as InputCell[], outputs: [] as OutputCell[] } as CKBVirtualResult['ckbRawTx'],
        commitment,
        sumInputsCapacity: '1000',
        needPaymasterCell: false,
      },
    };
    // FIXME: mock electrs getTransaction
    const btcTx = await cradle.electrs.getTransaction(transactionRequest.txid);
    const isValid = await transactionManager.verifyTransaction(transactionRequest, btcTx);
    expect(isValid).toBe(false);
  });

  test('verifyTransaction: should throw DelayedError for unconfirmed transaction', async () => {
    vi.spyOn(
      transactionManager as unknown as {
        getCommitmentFromBtcTx: (txid: string) => Promise<Buffer>;
      },
      'getCommitmentFromBtcTx',
    ).mockResolvedValueOnce(Buffer.from(commitment, 'hex'));
    vi.spyOn(transactionManager['cradle']['electrs'], 'getTransaction').mockResolvedValueOnce({
      status: { confirmed: false, block_height: 0 },
    } as unknown as Transaction);

    const transactionRequest: ITransactionRequest = {
      txid: 'bb8c92f11920824db22b379c0ef491dea2d819e721d5df296bebc67a0568ea0f',
      ckbVirtualResult: {
        ckbRawTx: { inputs: [] as InputCell[], outputs: [] as OutputCell[] } as CKBVirtualResult['ckbRawTx'],
        commitment,
        sumInputsCapacity: '1000',
        needPaymasterCell: false,
      },
    };

    // FIXME: mock electrs getTransaction
    const btcTx = await cradle.electrs.getTransaction(transactionRequest.txid);
    await expect(
      transactionManager.verifyTransaction(transactionRequest, btcTx),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test('enqueueTransaction: should be add transaction request to queue', async () => {
    const transactionRequest: ITransactionRequest = {
      txid: '0x123',
      ckbVirtualResult: {
        ckbRawTx: {} as CKBVirtualResult['ckbRawTx'],
        commitment: '0x123',
        sumInputsCapacity: '1000',
        needPaymasterCell: false,
      },
    };

    transactionManager.enqueueTransaction(transactionRequest);
    const count = await transactionManager['queue'].getJobCounts();
    const job = await transactionManager['queue'].getJob(transactionRequest.txid);
    expect(count.delayed).toBe(1);
    expect(job?.delay).toBe(cradle.env.TRANSACTION_QUEUE_JOB_DELAY);
  });
});
