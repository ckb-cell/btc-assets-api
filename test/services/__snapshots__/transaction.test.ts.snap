// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transactionManager > verifyTransaction: should throw DelayedError for unconfirmed transaction 1`] = `[DelayedError]`;
